### PR TITLE
Feature request: torch.broadcast_to

### PR DESCRIPTION
**Pull request by boeddeker**
I would like to have `broadcast_to`. Numpy has also a `broadcast_to` (https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.broadcast_to.html#numpy.broadcast_to).
`torch.broadcast_tensors` can solve some cases, but `broadcast_to` can solve more cases.

This PR is a suggestion for a python implementation.

Related to https://github.com/pytorch/pytorch/issues/8076 (`broadcast_tensors`).

### CODE REVIEW

**Review by zou3519 — CHANGES_REQUESTED** [review]
Thanks for the pull request, @boeddeker.

I had a couple of comments:
- We need to add an `autofunction` to https://raw.githubusercontent.com/pytorch/pytorch/master/docs/source/torch.rst for broadcast_to to show up in the documentatation
- Please add a test or two in test/test_torch.py
- It would be better if we could implement this as a native function in our C++ backend. I'm not sure if this is a blocker for this pull request, but it would be nice to have.

File: torch/functional.py [hunk 1]
```diff
@@ -55,6 +56,29 @@ def broadcast_tensors(*tensors):
...
+        torch.Size([4, 3])
+        >>> broadcast_to(torch.ones(1, 3), (4, 3)).shape
+        torch.Size([4, 3])
+    """
+    delta_len = tensor.dim() - len(shape)
+    if delta_len > 0:
+        # add singleton dimensions
+        tensor = tensor[[slice(None)] * delta_len]
+
+    return tensor.expand(shape)
```
Comment (vishwakftw) [inline torch/functional.py]: This function would never error out. For example:

```python
torch.broadcast_to(torch.randn(3), (3, 4))
```
would not be a valid broadcast.

I think using `torch._C._infer_size` here would be a good idea, as it errors out if the sizes are not broadcastable.

**Comment by SsnL** [comment]
I'm not sure I understand the differences in use cases of this and of `expand`

**Comment by boeddeker** [comment]
Sorry, my fault. I was not carefully reading the documentation. Forget this PR, `expand` supports adding dimensions.

Maybe adding 
```python
>>> torch.ones(3).expand(4, 3).shape
torch.Size([4, 3])
```
to the examples in the `expand` documentation would prevent the next one to do the same mistake.

When I searched for `broadcast_to` I found only `broadcast_tensors` in pytorch. Maybe a hint in `broadcast_tensors` to `expand` as a `broadcast_to` alternative would be usefull.
Or is renaming an option for numpy compatibility? (Or adding an alias)

**Comment by zou3519** [comment]
cc @umanwizard for the numpy compatibility ideas. I'll close this PR, but feel free to keep discussing

**Pull request closed without merging**